docs/spec/connection-handshake/L1_2.md

```diff
@@ -0,0 +1,626 @@
+  setConnection(local.connectionIdentifier, connectionEnd)
+}
+```
+
+
+Preconditions:
+  - The two proofs `remoteConnectionProof` and `remoteClientProof` should be correct; this is necessary in connection to properties [ICS3-Proto-2-ConnectionIntegrity] and [ICS3-Proto-3-StateConsistency].
+  - Validation of `local` party must pass, and the module running this handler must support at least one of the versions supplied in the input list `remoteVersions`.
+
+Postconditions:
+  - The module executing this handler stores a `ConnectionEnd` (matching with the arguments `local` and `remote`) in state `TRYOPEN`.
+  - No output; may abort if preconditions not met.
+
+
+###### ConnAckHandler
+
+Step 3.b.
+
+```golang
+func ConnAckHandler(local ConnectionParty,
+                    remoteVersion String,
+                    proofsHeight uint64,
+                    remoteConnectionProof ConnProof,
+                    remoteClientProof ClientProof)
+{
+  // Search the details of this connection in local store.
+  connectionEnd = getConnection(local.connectionIdentifier)
+
+  // This verification helps guarantee uniqueness and integrity.
+  abortTransactionUnless(connectionEnd != nil && connectionEnd.local == local)
+
+  // Another integrity check: 
+  // Should not overwrite a connection that is already `OPEN`.
+  abortTransactionUnless(connectionEnd.state == INIT
+                         || connectionEnd.state == TRYOPEN)
+
+  expectedConn = NewConnectionEnd('TRYOPEN', remote, local)
+  expectedClient = consensusState(remoteClientProof.height)
+
+  // Proofs verification.
+  localClientState = clientConsensusState(local.clientIdentifier, proofsHeight)
+  abortTransactionUnless(localClientState != nil)
+
+  // Connection proof specific verification.
+  abortTransactionUnless(verifyProof(
+      localClientState, remoteConnectionProof,
+      connectionPath(local.connectionIdentifier), expectedConn))
+
+  // Client proof verification.
+  abortTransactionUnless(remoteClientProof.height <= getCurrentHeight())
+  abortTransactionUnless(verifyProof(localClientState, remoteClientProof,
+        clientPath(local.clientIdentifier, remoteClientProof.height),
+        expectedClient))
+
+  // Termination (partial -- only at this end)
+  connectionEnd.state = OPEN
+  connectionEnd.version = remoteVersion
+
+  // Update the connectionEnd in local state.
+  setConnection(local.connectionIdentifier, connectionEnd)
+}
+```
+
+
+Preconditions:
+  - The module executing this handler already has a `ConnectionEnd` stored locally, matching the `local` argument, and in state `INIT` or `TRYOPEN`.
+  - The `local` argument is valid (syntactic check), and the `remoteVersion` is supported, i.e., compatible with the local chain -- see `getCompatibleVersions()`
+  - The two proofs `remoteConnectionProof` and `remoteClientProof` are syntactically and semantically correct; for guaranteeing properties [ICS3-Proto-2-ConnectionIntegrity] and [ICS3-Proto-3-StateConsistency]
```

Comment:
Why isn't this in the `tryOpen` also? I also want to make sure it is clear what syntactically correct means (non-null proofs? non-zero height?) and semantically correct (the proofs verify?)

Reply:
I removed the confusion around syntactic & semantic checks.